plugins/usbpro/OpenDeckDevice.cpp

```diff
@@ -0,0 +1,239 @@
+    UsbSerialDevice(owner, name, widget),
+    m_opendeck_widget(widget),
+    m_serial(),
+    m_got_parameters(false),
+    m_break_time(0),
+    m_mab_time(0),
+    m_rate(0) {
+  // Serial number is stored as series of 4 bytes.
+  // Display the serial in hex format so that it
+  // matches (partially) with USB serial number.
+  std::ostringstream str;
+  str << ToHex(serial);
+  m_serial = str.str();
+  str.str("");
+
+  // Firmware version is stored in major.minor format.
+  // This scheme uses the following encoding:
+  // MMMMMMMM mmmmmmmm
+  // M: major
+  // m: minor
+  uint8_t major;
+  uint8_t minor;
+
+  SplitUInt16(firmware_version, &major, &minor);
+
+  str << "Serial #: " << m_serial << ", firmware "
+      << static_cast<int>(major) << "." << static_cast<int>(minor);
+
+  m_opendeck_widget->GetParameters(NewSingleCallback(
+    this,
+    &OpenDeckDevice::UpdateParams));
+
+  // add the output port
+  OutputPort *output_port = new OpenDeckOutputPort(
+      this,
+      m_opendeck_widget,
+      0,
+      str.str(),
+      plugin_adaptor->WakeUpTime(),
+      5,  // allow up to 5 burst frames
+      fps_limit);
+  AddPort(output_port);
+}
+
+
+/*
+ * Stop this device
+ */
+void OpenDeckDevice::PrePortStop() {
+  m_opendeck_widget->Stop();
+}
+
+
+/*
+ * Handle device config messages
+ * @param controller An RpcController
+ * @param request the request data
+ * @param response the response to return
+ * @param done the closure to call once the request is complete
+ */
+void OpenDeckDevice::Configure(RpcController *controller,
+                               const string &request,
+                               string *response,
+                               ConfigureCallback *done) {
+  Request request_pb;
+  if (!request_pb.ParseFromString(request)) {
+    controller->SetFailed("Invalid Request");
+    done->Run();
+    return;
+  }
+
+  switch (request_pb.type()) {
+    case ola::plugin::usbpro::Request::USBPRO_PARAMETER_REQUEST:
+      HandleParametersRequest(controller, &request_pb, response, done);
+      break;
+    case ola::plugin::usbpro::Request::USBPRO_SERIAL_REQUEST:
+      HandleSerialRequest(controller, &request_pb, response, done);
+      break;
+    default:
+      controller->SetFailed("Invalid Request");
+      done->Run();
+  }
+}
+
+
+/**
+ * Update the cached param values
+ */
+void OpenDeckDevice::UpdateParams(bool status,
+                                  const usb_pro_parameters &params) {
+  if (status) {
+    m_got_parameters = true;
+    m_break_time = params.break_time;
+    m_mab_time = params.mab_time;
+    m_rate = params.rate;
+  }
+}
+
+
+/*
+ * Handle a parameter request. This may set some parameters in the widget.
+ * If no parameters are set we simply fetch the parameters and return them to
+ * the client. If we are setting parameters, we send a SetParam() request and
+ * then another GetParam() request in order to return the latest values to the
+ * client.
+ */
```

Comment:
This comment is probably stale now isn't it. Maybe replace it with one briefly explaining the technical issues why you can't change your timing?

Reply:
That's unrelated. The rate limiting is internal to the plugin, whereas MAB and Break (in Enttec land) are set on the device and define the actual output from the on-board micro.

Reply:
Ah I see, thanks!